tutorials/dataframe/df107_SingleTopAnalysis.py

```diff
@@ -0,0 +1,245 @@
+    const ROOT::Math::PtEtaPhiEVector bjet(jet_pt / 1000.0, jet_eta, jet_phi, jet_e / 1000.0);
+    // Please note that we treat here the missing transverse energy as the neutrino, even though the z compontent is missing!
+    return (lep + met + bjet).M();
+}
+""")
+
+histos = {}
+for s in samples:
+    df[s] = df[s].Define("top_mass", "ComputeTopMass(lep_pt[idx_lep], lep_eta[idx_lep], lep_phi[idx_lep], lep_E[idx_lep], jet_pt[idx_tagged], jet_eta[idx_tagged], jet_phi[idx_tagged], jet_E[idx_tagged], met_et, met_phi)")
+    histos[s] = df[s].Histo1D(ROOT.RDF.TH1DModel("top_mass", "", 10, 100, 300), "top_mass", "weight")
+
+# Run the event loop and merge histograms of the respective processes
+
+def merge_histos(label):
+    h = None
+    for i, d in enumerate(files[label]):
+        t = histos[d[1]].GetValue()
+        if i == 0: h = t.Clone()
+        else: h.Add(t)
+    h.SetNameTitle(label, label)
+    return h
+
+data = merge_histos("data")
+twtb = merge_histos("twtb")
+singletop = merge_histos("singletop")
+wjets = merge_histos("wjets")
+
+# Create the plot
+
+# Set styles
+ROOT.gROOT.SetStyle("ATLAS")
+
+# Create canvas with pad
+c = ROOT.TCanvas("c", "", 600, 600)
+pad = ROOT.TPad("upper_pad", "", 0, 0, 1, 1)
+pad.SetTickx(False)
+pad.SetTicky(False)
+pad.Draw()
+pad.cd()
+
+# Draw stack with MC contributions
+stack = ROOT.THStack()
+wjets.Scale(1.1) # Corrected normalization derived from a validation region
+for h, color in zip(
+        [wjets, twtb, singletop],
+        [(222, 90, 106), (155, 152, 204), (208, 240, 193)]):
+    h.SetLineWidth(1)
+    h.SetLineColor(1)
+    h.SetFillColor(ROOT.TColor.GetColor(*color))
+    stack.Add(h)
+stack.Draw("HIST")
+stack.GetXaxis().SetTitle("m_{W(l#nu)+b} [GeV]")
+stack.GetYaxis().SetTitle("Events")
+stack.GetYaxis().SetLabelSize(0.04)
+stack.GetYaxis().SetTitleSize(0.045)
+stack.GetXaxis().SetLabelSize(0.04)
+stack.GetXaxis().SetTitleSize(0.045)
+stack.SetMinimum(0)
+stack.SetMaximum(4000 * lumi_scale)
+stack.GetYaxis().ChangeLabel(1, -1, 0)
+
+# Draw data
+data.SetMarkerStyle(20)
+data.SetMarkerSize(1.2)
+data.SetLineWidth(2)
+data.SetLineColor(ROOT.kBlack)
+data.Draw("E SAME")
+
+# Add legend
+legend = ROOT.TLegend(0.60, 0.65, 0.92, 0.92)
+legend.SetTextFont(42)
+legend.SetFillStyle(0)
+legend.SetBorderSize(0)
+legend.SetTextSize(0.035)
+legend.SetTextAlign(32)
+legend.AddEntry(data, "Data" ,"lep")
+legend.AddEntry(singletop, "Single top + jet", "f")
+legend.AddEntry(twtb, "t#bar{t},Wt,t#bar{b}", "f")
+legend.AddEntry(wjets, "W+jets", "f")
+legend.Draw("SAME")
+
+# Add ATLAS label
+text = ROOT.TLatex()
+text.SetNDC()
+text.SetTextFont(72)
+text.SetTextSize(0.045)
+text.DrawLatex(0.21, 0.86, "ATLAS")
+text.SetTextFont(42)
+text.DrawLatex(0.21 + 0.16, 0.86, "Open Data")
+text.SetTextSize(0.04)
+text.DrawLatex(0.21, 0.80, "#sqrt{{s}} = 13 TeV, {:.1f} fb^{{-1}}".format(lumi * lumi_scale / 1000.0))
+
+# Save the plot
+c.SaveAs("SingleTopAnalysis.pdf")
```

Comment:
Possibly wrap the analysis and the plotting in separate functions and add `if __name__ == "__main__"` as in df103 ?

Reply:
hmm, we haven't had this in the last ones (df105 and df106), I think mainly because the script is simple enough not to have the main. Also, mains in Python are rather if you don't plan to import the file in another script.

Reply:
Ah wait you meant factorizing the RDF part and the plotting part! Sry misunderstood you :)

I'm not sure about this, having a linear flow is more simpler since that's how people think through something. Although from the programming point of view factorizing is nicer :/

Reply:
I think I'm against using functions here since we don't reuse the code. But feel free to convince me!

Reply:
I'm never sure if I like the forceful saving of plots into files. When users do this in forum examples, I'm always annoyed that I have to delete it now. Also, when running interactively, it's nice if you can just see the plot immediately, so a `c.Draw()` could be nice.

Reply:
`c.Draw()` does not work so well in PyROOT, see https://sft.its.cern.ch/jira/browse/ROOT-10960